Standardize typing in TaskSampler.episodic_collate_fn()

Aml-Hassan-Abd-El-hamid opened the pull request:
Hi @ebennequin ,
I added a code to check the type of input of the method episodic_collate_fn and raise a TypeError if the input wasn't a List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]].
If the input was a List[Tuple[Tensor, 0-dim Tensor]], a for loop would change that to be List[Tuple[Tensor, int]].
I tested the code after those changes and the error from that [issue #89](https://github.com/sicara/easy-few-shot-learning/issues/89) was gone, you can see the test [here](https://www.kaggle.com/code/amlhassan/rn-with-binary-classification-first-error)
If you'd like me to change anything about that code, please let me know.

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -80,6 +80,33 @@ def episodic_collate_fn(
                 - their labels,
                 - the dataset class ids of the class sampled in the episode
         """
+        #the input_data should be a list
```
(!) I don't think we need to check that `input_data` is a list as it is a construction of `torch.utils.data.DataLoader` which we can trust. Plus, if it's not a list, it will break very fast so it's not a huge risk.

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -80,6 +80,33 @@ def episodic_collate_fn(
+                "check out the type of the output of the .getitem() method of your dataset and make sure it's"
+                "a List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]]."
+            )
+        #the instances inside the input_data list should be all Tuples
```
(!) Same here, we don't need to check that each instance is a tuple

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -80,6 +80,33 @@ def episodic_collate_fn(
                 - their labels,
                 - the dataset class ids of the class sampled in the episode
         """
```
(!) If we're going to accept labels of type `Tensor` we need to change the method's signature to 

```python
def episodic_collate_fn(
        self, input_data: List[Tuple[Tensor, Union[Tensor, int]]]
    ) -> Tuple[Tensor, Tensor, Tensor, Tensor, List[int]]:
```

ebennequin replied:
Also the expected input and raised errors must be specified in docstring

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -80,6 +80,33 @@ def episodic_collate_fn(
+                "check out the type of the output of the .getitem() method of your dataset and make sure it's" 
+                "a List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]]."
+            )
+        #inside the tuple there should be  List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]]
```
(!) This need to be moved to a new method

- lightens the core `episodic_collate_fn()` method
- allows to specify a clear docstring instead of inline comments

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -80,6 +80,33 @@ def episodic_collate_fn(
+                "a List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]]."
+            )
+        #inside the tuple there should be  List[Tuple[Tensor, int]] or List[Tuple[Tensor, 0-dim Tensor]]
+        if not all( isinstance(n[0],torch.Tensor) and isinstance(n[1],int) for n in input_data):
```
(!) Please use explicit names for your variables to make the code easily understandable. Here I'd suggest something like 

```python
for image, label in input_data:
	# 1. Check that image is a tensor. If not, raise a specific TypeError to say that the image must be a tensor and not type(image)
	# 2. Check that label is an int. If not, check that it is a tensor. If it's neither, raise TypeError.
	# 3. Check that label is a 0-dim tensor. If not, raise a ValueError. If so, cast into int.
```

Aml-Hassan-Abd-El-hamid replied:
Hey @ebennequin , thank you for your notes, I'm still learning so it's really helpful to hear feedback for my code, I think I'll have enough time to change my PR tomorrow or something, I hope that's ok with you.

ebennequin reviewed (changes requested):
Thank you so much for contributing! I have raised a few comments, if you feel like updating your PR that's great! I'd understand if you don't have time, just let me know.

Aml-Hassan-Abd-El-hamid commented:
Hi @ebennequin:
I did the following:
- I removed the list and the tuple check.
- I changed the method's signature.
- I specified the expected input and raised errors in the docstring.
- I created a method to check the input of the episodic_collate_fn method and return 2 bools: 1 to indicate whether the input is the right type or not and the other to indicate whether the input label where tensors or not.
- Also created a method to turn 0-dim tensor labels into int labels.
-  And tried to use explicit names for your variables as much as I could :)

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -1,19 +1,14 @@
 import random
-from typing import List, Tuple, Iterator
-
+from typing import List, Tuple, Iterator,Union
```
Make sure that you do it after all your other changes are made ;)

Aml-Hassan-Abd-El-hamid replied:
Hey @ebennequin, I did the other changes, but I can't find a file called black easyfsl, I think that I don't understand what you mean exactly by that, so could you please elaborate more?

ebennequin replied:
[Black](https://pypi.org/project/black/) is a very useful tool to format python code. After installing `black` (it is in the `dev_requirements.txt` so it should already be installed), you can call `black easyfsl` to format all python files inside the `easyfsl` folder.

Aml-Hassan-Abd-El-hamid replied:
Thank you very much for that information @ebennequin , It's my first time hearing about Black, so that was really helpful :)

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -63,26 +56,37 @@ def __iter__(self) -> Iterator[List[int]]:
-
+        #checking the input type
+        right_input_type, o_dim_tesors_labesl= self.check_episodic_collate_fn_input(input_data)
+        if not right_input_type:
```
As I see it `check_episodic_collate_fn_input()` could be reframed as: `cast_input_labels_to_int`, take `input_data` as argument and output an updated `input_data` with all labels as integers

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -92,18 +96,54 @@ def episodic_collate_fn(
+                - True if the input labels needs to be transformed from tensor to int and False otherwise.
+        """
+        #that flag is true when the input labels are  0-dim tensors
+        o_dim_tesors_labesl=False
```
Most IDEs have (or allow plugins with) automatic spellchecks that spot such spelling errors in variable names, perhaps it could help you.

ebennequin replied:
Also I think it's bad practice to use a `o` to mean `0`. This variable could be named `labels_are_0_dim_tensors`
(but as I suggest in a later comment I don't think we need the variable)

ebennequin commented on easyfsl/samplers/task_sampler.py:
```diff
@@ -92,18 +96,54 @@ def episodic_collate_fn(
+        for img, label in input_data:
+            #if that label isn't int then check if it's a 0-tensor
+            if not  (isinstance(img,torch.Tensor) and isinstance(label,int)):
+                if not (isinstance(img,torch.Tensor) and isinstance(label,torch.Tensor) and label.ndim==0):
```
(!) You're checking `img`'s type twice. I suggest a cleaner case switch, something like this:

```python
for item_index, (image, label) in enumerate(input_data):
	if not isinstance(image, Tensor): 
		raise TypeError("...")
	if not isinstance(label, int):
		if not isinstance(label, Tensor):
			raise TypeError("...")
		if label.ndim !=0:
			raise ValueError("...")
		input_data[item_index] = (image, int(label))
```

ebennequin reviewed (changes requested):
Thank for the updates! There still are a few things I think should be done differently, so I left you some comments.

Sorry about the extra work I am asking of you on this, I hope this won't be too inconvenient. I'm really grateful for your contribution!

ebennequin commented:
Thank you for your great work! By the end of the week I will just:

- apply standard formatting
- write some tests to ensure our changes are robust
- merge
- release a new minor version of easyfsl including these changes.

Thank you again!

AntoineToubhans reviewed (approved):
Great work, thank you guys!
